### PR TITLE
feat(lit-api-server): language registry + /get_supported_languages discovery (CPL-349 phase 1)

### PR DESCRIPTION
**Pull request by GTC6244**
Phase 1 of [CPL-349](https://linear.app/litprotocol/issue/CPL-349/cvm-language-support) (design: #561). Adds the language capability surface only — no execution-path changes; the gVisor-dependent phases stack on #557/#558.

## What's here

- **`lit-api-server/src/actions/languages.rs`** — the registry that will drive both discovery and request admission for multi-language actions: `LanguageFeature`/`LanguageRuntime`/`ExecutionMethod`/`ExecutionModel`, a built-in table of known languages (javascript→deno; python→gvisor with `python3.13`/`python3.12`; rust→gvisor), and the `LIT_SUPPORTED_LANGUAGES` allowlist parser.
- **Allowlist grammar** (the design doc left the encoding TBD): entries `;`-separated, each `lang[:runtime[:runtime...]]|method[,method...]`, e.g. `javascript|raw_script; python:python3.13:python3.12|raw_script,bundle; rust|bundle`. First runtime listed is the default; repeated `|` are tolerated so the doc's example strings parse as-is. `Display` emits the canonical form and round-trips.
- **Startup wiring (`main.rs`)** — parsed once before the restart loop into `Arc<SupportedLanguages>` managed state. Unset/blank env falls back to today's surface (`javascript|raw_script`); a set-but-invalid value fails the boot, so a deploy-config typo can't misadvertise capability.
- **`GET /get_supported_languages`** in `endpoints/configuration.rs`, unauthenticated, mirroring `get_lit_action_client_config`; `SupportedLanguagesResponse` in `models/response.rs`.
- **k6 client + spec regenerated** (`spec.json`, `k6/litApiServer.ts`) for the k6-client-check gate.
- **Docs**: `docs/lit-actions/languages.mdx` (capability discovery), linked in the nav.

## Notes for review

- `prewarmed` is hardcoded `false` until the install cache (phase 2) and deploy pre-warm (phase 5) exist — honest advertising, wired later.
- `allows(language, method)` is the admission hook `/lit_raw_action` (phase 3) and `/lit_binary_action` will call; unused by routes in this PR.
- Runtime versions pinned in the known table (`python3.13`→3.13.1, `python3.12`→3.12.7) per the design doc; these become install recipes in phase 2.

## Tests

- Parser unit tests: round-trip of the canonical encoding, the design doc's `next` allowlist, unknown language/runtime/method rejection, duplicate/empty rejection.
- Rocket local-client test dispatching `GET /get_supported_languages` against managed state and asserting the JSON payload.
- `cargo +1.91 fmt --check`, `clippy --all-targets --all-features` (0 warnings), full crate test suite green (280 tests).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by GTC6244** [comment]
The four red Rust jobs fail only at the `cargo deny` step — two RustSec advisories published today (RUSTSEC-2026-0204/0205) that break every branch, including the docs-only #561. Fix is #563 (green across the board). Once it merges, re-running the failed jobs here picks it up — this branch touches no lockfiles. Everything else (fmt, clippy, build, tests, k6-client-check) is green.

File: lit-api-server/src/actions/languages.rs [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+                if runtimes.iter().any(|r| r.id == id) {
+                    bail!("runtime {id:?} listed more than once for language {name:?}");
+                }
+                runtimes.push(LanguageRuntime {
+                    id: id.to_string(),
+                    version: version.to_string(),
+                    is_default: runtimes.is_empty(),
+                    prewarmed: false,
+                });
+            }
```
[the remark this answers is not in the document]
Comment (GTC6244) [reply lit-api-server/src/actions/languages.rs:214]: Good catch — fixed in ec168e6a.

A runtime-bearing language listed without an explicit runtime list (e.g. `python|raw_script`) now expands to **all runtimes this build knows about**, with the first (`python3.13`) as the default — so the discovery payload never advertises a runtime-bearing language with an empty `runtimes` array or no default.

Decision detail: I went with default-to-all rather than rejecting the entry, because it makes the optional `[:runtime...]` grammar coherent and forgiving on `next` (permissive), while explicit listing remains the mechanism to restrict a subset — which is exactly what prod does (`python:python3.13|bundle`). Languages with no known runtimes (JS, Rust) correctly stay empty. Added `runtime_bearing_language_without_listed_runtimes_gets_all_known` and `runtime_less_language_without_runtimes_stays_empty` tests.

File: docs/lit-actions/languages.mdx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      "runtimes": [],
+      "methods": ["raw_script"]
+    },
+    {
+      "name": "python",
+      "display_name": "Python",
+      "execution_model": "gvisor",
+      "runtimes": [
+        { "id": "python3.13", "version": "3.13.1", "is_default": true, "prewarmed": true },
+        { "id": "python3.12", "version": "3.12.7", "is_default": false, "prewarmed": true }
```
[the remark this answers is not in the document]
Comment (GTC6244) [reply docs/lit-actions/languages.mdx]: Fixed in ec168e6a. The example now shows `prewarmed: false` for both Python runtimes, matching the registry's hardcoded value for this phase, and I added a line under the example noting that `prewarmed` is `false` for every runtime until pre-warmed cache profiles land in a later phase (cross-referencing the existing rollout note).

**Comment by clawdbot-glitch003** [comment]
## 🤖 Code review (Claude)

**Summary:** Phase 1 of CPL-349: adds a language registry (`lit-api-server/src/actions/languages.rs`) with a built-in table of known languages, a `LIT_SUPPORTED_LANGUAGES` allowlist parser that fail-fast validates deploy config at boot, an unauthenticated `GET /get_supported_languages` discovery endpoint, regenerated `spec.json`/k6 client, and a docs page. No execution-path changes; `allows()` is the future admission hook.

I verified against the local codebase: the docs curl path `/core/v1/get_supported_languages` matches the core-route mount in `main.rs` (`.mount("/core/v1/", core_routes)`); the `/lit-actions/index`-style doc links match existing convention (`docs/quickstart.mdx`, `docs/index.mdx`); `std::process::exit(1)` on bad config matches the existing Stripe-init fail-fast pattern; and `k6/litApiServer.ts` drift is gated by `.github/workflows/k6-client-check.yml`, so the hand-in-sync risk is covered by CI.

### Findings

1. **Implicit "all runtimes" expansion can widen prod capability on binary upgrade** — `languages.rs` (`parse`, the `if runtimes.is_empty()` block, ~L455). `python|raw_script` expands to *every* runtime in `KNOWN_LANGUAGES` at build time. If a later release adds `python3.14` to the known table, any node whose allowlist omits explicit runtimes silently starts advertising (and, once phase 3 lands, admitting) it — capability drift with no config change and no boot-time signal. The comment documents the behavior and the PR notes prod should list runtimes explicitly, but nothing enforces that. Consider requiring explicit runtimes for runtime-bearing languages (or at least logging the expansion at startup) so the allowlist stays a true allowlist. Not a blocker for phase 1 since nothing executes yet, but worth deciding before phase 3 wires admission.

2. **`allows()` has no runtime dimension** — `languages.rs` L507. Admission in phase 3 needs to validate the requested `runtime` id against the advertised set (and resolve the default when omitted), but the registry exposes no `allows_runtime`/`default_runtime` helper. Fine to defer, just flagging so the phase-3 PR doesn't check only `(language, method)` and accept an unadvertised runtime string.

3. **Misleading parse error for a missing language** — `languages.rs` L411-417. Because empty `|`-fields are filtered before positional assignment, `"|raw_script"` parses `raw_script` as the *language spec* is skipped… actually it becomes the spec, and the entry fails with `missing methods` rather than "unknown language"/"missing language". Same mechanism means a trailing pipe (`javascript|raw_script|`) is silently tolerated. Both are rejected-or-harmless, so this is only an error-message-quality issue.

### Nits

- `docs/lit-actions/languages.mdx` L90 (`prewarmed` row): "production only admits pre-warmed runtimes" is stated as present-tense fact, but nothing enforces it yet (`prewarmed` is hardcoded `false` and admission isn't wired). The closing `<Note>` covers the phasing, but this row reads as current behavior — suggest "will only admit" or moving the claim into the note.
- `languages.rs`: `ExecutionMethod::parse`/`as_str` duplicate the serde `snake_case` renames by hand. A `serde_plain`/`serde_json::from_value` round-trip (or `strum`) would keep them from drifting if a variant is added, though with three variants and tests this is fine as-is.
- The doc comment says "Repeated `|` are tolerated" — the round-trip test covers `javascript||raw_script`, good; consider one test asserting the trailing-`|` case so the tolerance is pinned intentionally.

Test coverage is solid (parser edge cases, round-trip, Rocket local-client dispatch against managed state). The endpoint being unauthenticated matches the `get_lit_action_client_config` precedent and only exposes deploy config, which is acceptable.

**Verdict: LGTM with nits.** Nothing blocking — finding 1 is a design decision to settle before phase 3 rather than a bug in this PR.

**Comment by GTC6244** [comment]
Thanks for the review — went through each item. Addressed in `d87e559d`.

**Findings**

1. **Implicit "all runtimes" expansion widens capability on binary upgrade** — good catch; this is the tension from the earlier Copilot fix (which wanted `python|raw_script` to not advertise empty `runtimes`/no default). I kept default-to-all (ergonomic on `next`, always yields a default) but took your fallback suggestion: the expansion now emits a startup `tracing::info!` naming the language and the runtimes it defaulted to, so a widening after a `KNOWN_LANGUAGES` addition is visible without a config diff. Prod pins runtimes explicitly and never hits the path. The stronger "require explicit runtimes" option is a real alternative — I've left it as a documented decision to settle in the phase-3 admission PR rather than reverse the just-merged behavior now.

2. **`allows()` has no runtime dimension** — agreed, deferred. I didn't add `allows_runtime`/`default_runtime` now because there's no caller yet and CI runs `clippy -D warnings`, so the helpers would fail as dead code. Flagged for the phase-3 PR so admission validates the `runtime` id (and resolves the default), not just `(language, method)`.

3. **Misleading parse error for missing language / tolerated trailing `|`** — left the behavior as-is: the empty-`|`-field filter is what lets the design-doc's `javascript||raw_script` parse, and removing it would turn that into a "more than two fields" error. All outcomes are still correct rejections/acceptances (error-message quality only, as you noted). I did pin the trailing-/repeated-`|` tolerance with a new test so it's intentional.

**Nits**

- **`prewarmed` doc row present-tense** — fixed; reworded to "will only admit …" and noted it's currently `false` for every runtime.
- **Hand-written `ExecutionMethod::parse`/`as_str` mirroring serde** — left as-is per your own "fine as-is with three variants + tests"; not worth a `strum`/`serde_plain` dependency here. The methods are covered by the unknown-token test.
- **Trailing-`|` test** — added (`tolerates_repeated_and_trailing_pipes`).

No API-surface change, so `spec.json`/k6 didn't need regenerating; fmt/clippy/tests green locally.
